File: src/pipeline.ts
import type { ExecutorFailureClass, RunConfig, TicketClass, TicketSpec, WorkerName } from './types.js';
import type { ReadinessGateResult } from './validation/readiness-gate.js';
import type { RunLogger } from './logger.js';
import type { ImplementationPlan } from './planner.js';
import { checkSpec, planFromTicket } from './planner.js';

/**
 * Pipeline state machine (PRD section 10).
 * fetch -> plan -> spec-check -> implement -> validate(G3) -> publish.
 * Sandbox-dependent gates (G1/G2) are injected by the caller when available.
 */

export type StageOutcome =
  | { stage: 'clarify'; question: string }
  | { stage: 'plan'; summary: string; tasks: string[] }
  | {
      stage: 'implement';
      worker: WorkerName;
      worktreePath?: string;
      branch?: string;
      attempts: number;
      ok: boolean;
    }
  | { stage: 'validate'; passed: boolean }
  | { stage: 'publish'; prUrl?: string; note: string }
  | { stage: 'failed'; reason: string };

export interface ImplementResult {
  ok: boolean;
  worker: WorkerName;
  worktreePath?: string;
  branch?: string;
  attempts: number;
  /** Executor failure class on failure (PRD:775 / Q24 taxonomy mirror). */
  failureClass?: ExecutorFailureClass;
}

export interface PipelineDeps {
  fetchTicket(ticketId: string): Promise<TicketSpec>;
  postTicketComment?(trackerInternalId: string, comment: string): Promise<void>;
  /** G0 issue-readiness gate: type-specific ready-for-dev scoring before worker dispatch. Optional so hand-built test deps stay valid; buildDeps wires the deterministic default. */
  runGateG0?(ticket: TicketSpec, classification: TicketClass): ReadinessGateResult;
  runGateG3(repoPath: string, classification: import('./types.js').TicketClass): {
    passed: boolean;
    findings: unknown[];
    detail?: string;
  };
  /** Real worker dispatch; injected so the pipeline stays unit-testable. */
  implementStage?(cfg: RunConfig, plan: ImplementationPlan, log: RunLogger): Promise<ImplementResult>;
  /** G2 migration-apply gate; injected by the caller (Docker-based). */
  runGateG2?(worktreePath: string, timeoutMs: number): Promise<{ passed: boolean; skipped?: boolean; detail?: string }>;
  /** G4 async-review gate; injected by the caller (heuristic scan in cli/deps). */
  runGateG4?(worktreePath: string): Promise<{ passed: boolean; detail?: string }>;
  /** G1 test gate; injected by the caller (repo-native test run in cli.ts). */
  runGateG1?(worktreePath: string, timeoutMs: number): Promise<{ passed: boolean; skipped?: boolean; detail?: string }>;
  /** PR publishing; injected by the caller (gh-based in cli.ts). */
  publishStage?(cfg: RunConfig, plan: ImplementationPlan, impl: ImplementResult): Promise<string | undefined>;
}

export async function runPipeline(cfg: RunConfig, deps: PipelineDeps, log: RunLogger): Promise<StageOutcome[]> {
  const outcomes: StageOutcome[] = [];

  // Stage: fetch
  log.info('fetch', `Fetching ticket ${cfg.ticketId}`);
  const ticket = await deps.fetchTicket(cfg.ticketId);
  log.info('fetch', `Fetched "${ticket.title}"`, { labels: ticket.labels });

  // Stage: plan
  const plan = planFromTicket(ticket);
  log.info('plan', `Classified as ${plan.classification}`, { tasks: plan.tasks });

  // Gate G0 (issue readiness): type-specific ready-for-dev scoring BEFORE
  // worker dispatch. Rejected tickets get actionable findings (posted to the
  // tracker when possible) so credits are not burned on under-specified work.
  if (deps.runGateG0) {
    const g0 = deps.runGateG0(ticket, plan.classification);
    log.info('validate', `G0 ${g0.skipped ? 'skipped' : g0.passed ? 'passed' : 'rejected'}: ${(g0.detail ?? '').split('\n')[0]}`, {
      score: g0.score,
      threshold: g0.threshold,
    });
    outcomes.push({ stage: 'validate', passed: g0.passed });
    if (!g0.passed) {
      log.warn('clarify', 'G0 readiness gate rejected ticket', { findings: g0.findings.length });
      if (ticket.trackerInternalId && deps.postTicketComment) {
        try {
          await deps.postTicketComment(ticket.trackerInternalId, g0.detail ?? 'G0 readiness gate rejected this ticket');
        } catch (err) {
          log.warn('clarify', `Failed to post G0 rejection comment: ${(err as Error).message}`);
        }
      }
      outcomes.push({ stage: 'failed', reason: `G0 readiness gate rejected: ${g0.detail ?? 'score below threshold'}` });
      return outcomes;
    }
  }

  // Spec check (FR-TICKET-05)
  const spec = checkSpec(ticket);
  if (!spec.sufficient && spec.question) {
    log.warn('clarify', 'Insufficient specification', {});
    outcomes.push({ stage: 'clarify', question: spec.question });
    if (ticket.trackerInternalId && deps.postTicketComment) {
      try {
        await deps.postTicketComment(ticket.trackerInternalId, spec.question);
      } catch (err) {
        log.warn('clarify', `Failed to post clarification comment: ${(err as Error).message}`);
      }
    }
    return outcomes;
  }
  outcomes.push({ stage: 'plan', summary: plan.classification, tasks: plan.tasks });

  if (cfg.dryRun) {
    log.info('plan', 'Dry-run: stopping before implement', {});
    return outcomes;
  }

  // Stage: implement
  if (!deps.implementStage) {
    outcomes.push({ stage: 'failed', reason: 'no worker dispatch configured' });
    return outcomes;
  }
  const impl = await deps.implementStage(cfg, plan, log);
  outcomes.push({
    stage: 'implement',
    worker: impl.worker,
    worktreePath: impl.worktreePath,
    branch: impl.branch,
    attempts: impl.attempts,
    ok: impl.ok,
  });
  if (!impl.ok) {
    outcomes.push({ stage: 'failed', reason: 'worker failed to produce a diff within retry budget' });
    return outcomes;
  }

  // Stage: validate — G1 (tests) then G3 (migration static analysis)
  const gatePath = impl.worktreePath ?? cfg.repoPath;

  if (deps.runGateG1) {
    const g1 = await deps.runGateG1(gatePath, cfg.timeoutMs);
    log.info('validate', `G1 ${g1.skipped ? 'skipped' : g1.passed ? 'passed' : 'failed'}${g1.detail ? `: ${g1.detail.split('\n')[0]}` : ''}`, {});
    outcomes.push({ stage: 'validate', passed: g1.passed });
    if (!g1.passed) {
      outcomes.push({ stage: 'failed', reason: `test gate failed: ${g1.detail ?? 'no detail'}` });
      return outcomes;
    }
  }

  if (deps.runGateG2 && plan.classification === 'migration-required') {
    const g2 = await deps.runGateG2(gatePath, cfg.timeoutMs);
    log.info('validate', `G2 ${g2.skipped ? 'skipped' : g2.passed ? 'passed' : 'failed'}${g2.detail ? `: ${g2.detail.split(';')[0]}` : ''}`, {});
    outcomes.push({ stage: 'validate', passed: g2.passed });
    if (!g2.passed) {
      outcomes.push({ stage: 'failed', reason: `migration apply gate failed: ${g2.detail ?? 'no detail'}` });
      return outcomes;
    }
  }

  const g3 = deps.runGateG3(gatePath, plan.classification);
  log.info('validate', `G3 ${g3.passed ? 'passed' : 'failed'}: ${g3.detail ?? ''}`, {});
  outcomes.push({ stage: 'validate', passed: g3.passed });
  if (!g3.passed) {
    outcomes.push({ stage: 'failed', reason: 'migration static gate failed' });
    return outcomes;
  }

  // Stage: validate — G4 async review for consumer-classified tickets
  if (deps.runGateG4 && plan.classification === 'consumer-only') {
    const g4 = await deps.runGateG4(gatePath);
    log.info('validate', `G4 ${g4.passed ? 'passed' : 'failed'}: ${g4.detail ?? ''}`, {});
    outcomes.push({ stage: 'validate', passed: g4.passed });
    if (!g4.passed) {
      outcomes.push({ stage: 'failed', reason: `async review gate failed: ${g4.detail ?? 'no detail'}` });
      return outcomes;
    }
  }

  // Stage: publish — headless mode publishes; interactive mode defers to the caller
  if (cfg.autoPr && deps.publishStage) {
    try {
      const prUrl = await deps.publishStage(cfg, plan, impl);
      log.info('publish', `PR opened: ${prUrl ?? 'unknown'}`, {});
      outcomes.push({ stage: 'publish', prUrl, note: 'auto-pr enabled' });
    } catch (err) {
      log.error('publish', `PR creation failed: ${(err as Error).message}`);
      outcomes.push({ stage: 'failed', reason: `PR creation failed: ${(err as Error).message}` });
      return outcomes;
    }
  } else {
    outcomes.push({ stage: 'publish', note: cfg.autoPr ? 'no publisher configured' : 'awaiting approval' });
  }
  return outcomes;
}

File: internal/pipeline/pipeline.go
// Package pipeline is the Go port of src/pipeline.ts, src/task.ts,
// src/consume.ts (the parts not already in internal/gates), src/planner.ts,
// src/orchestrator/planner.ts, src/fleet.ts, src/create.ts, src/remote.ts,
// src/runregistry.ts and src/resilience/reaper.ts (FR-GO-07 remainder,
// issue #223): the ticket→PR pipeline state machine and the command-body
// helpers the cli wires.
//
// Byte-parity: error strings and printed shapes mirror the TypeScript
// originals exactly; tests pin them.
package pipeline

import (
	"fmt"
	"github.com/FreePeak/devagent/internal/gates"
	"github.com/FreePeak/devagent/internal/ledger"
	"github.com/FreePeak/devagent/internal/scout"
	"strings"
)

// RunLog mirrors the RunLogger surface the TS pipeline consumes
// (src/logger.ts). *ledger.RunLogger satisfies it; tests use a no-op.
type RunLog interface {
	Info(stage ledger.RunStage, message string, data []ledger.KV)
	Warn(stage ledger.RunStage, message string, data []ledger.KV)
	Error(stage ledger.RunStage, message string, data []ledger.KV)
}

// TicketClass mirrors types.ts TicketClass.
type TicketClass = scout.TicketClass

// WorkerName mirrors types.ts WorkerName.
type WorkerName = string

// ExecutorFailureClass mirrors types.ts ExecutorFailureClass.
type ExecutorFailureClass = string

// TicketSpec mirrors types.ts TicketSpec (via the scout port).
type TicketSpec = scout.TicketSpec

// ImplementationPlan mirrors planner.ts ImplementationPlan.
type ImplementationPlan = scout.ImplementationPlan

// CleanupMode mirrors the TS 'auto' | 'keep' | 'always' union.
type CleanupMode = string

const (
	CleanupAuto   CleanupMode = "auto"
	CleanupKeep   CleanupMode = "keep"
	CleanupAlways CleanupMode = "always"
)

// RunConfig mirrors pipeline.ts RunConfig (the cli.ts cfg object shape).
type RunConfig struct {
	TicketID          string
	RepoPath          string
	Worker            WorkerName
	Model             string // "" = unset
	Variant           string // "" = unset
	AutoMerge         bool
	AutoPr            bool
	Interactive       bool
	MaxLoops          int
	TimeoutMs         int
	DryRun            bool
	Cleanup           CleanupMode
	DropOrcaWorkspace bool
}

// StageOutcome mirrors the pipeline.ts StageOutcome union: one struct with a
// discriminating Stage field; only the fields the TS variant carries are set.
type StageOutcome struct {
	Stage        string     `json:"stage"` // clarify|plan|implement|validate|publish|failed
	Question     string     `json:"question,omitempty"`
	Summary      string     `json:"summary,omitempty"`
	Tasks        []string   `json:"tasks,omitempty"`
	Worker       WorkerName `json:"worker,omitempty"`
	WorktreePath string     `json:"worktreePath,omitempty"`
	Branch       string     `json:"branch,omitempty"`
	Attempts     int        `json:"attempts,omitempty"`
	OK           bool       `json:"ok,omitempty"`
	KgEvidence   any        `json:"kgEvidence,omitempty"`
	Passed       bool       `json:"passed,omitempty"`
	PRURL        string     `json:"prUrl,omitempty"`
	Note         string     `json:"note,omitempty"`
	Reason       string     `json:"reason,omitempty"`
}

// ImplementResult mirrors pipeline.ts ImplementResult.
type ImplementResult struct {
	OK           bool
	Worker       WorkerName
	WorktreePath string // "" = undefined
	Branch       string // "" = undefined
	Attempts     int
	FailureClass ExecutorFailureClass // "" = undefined
	KgEvidence   any                  // nil = undefined
}

// GateCheck mirrors the runGateG3 return shape ({passed, findings, detail?}).
type GateCheck struct {
	Passed   bool
	Findings []gates.Finding
	Detail   string
}

// G2G4Result mirrors the async gate result shape ({passed, skipped?, detail?}).
type G2G4Result struct {
	Passed  bool
	Skipped bool
	Detail  string
}

// ReadinessResult mirrors the runGateG0 return shape (validation/readiness-gate).
type ReadinessResult struct {
	Gate      string
	Passed    bool
	Skipped   bool
	Score     int
	Threshold int
	Findings  []gates.Finding
	Detail    string
}

// PipelineDeps mirrors pipeline.ts PipelineDeps. Optional hooks are nil.
type PipelineDeps struct {
	FetchTicket       func(ticketID string) (TicketSpec, error)
	PostTicketComment func(trackerInternalID, comment string) error
	// RunGateG0: issue-readiness gate before worker dispatch.
	RunGateG0 func(ticket TicketSpec, classification TicketClass) ReadinessResult
	RunGateG3 func(repoPath string, classification TicketClass) GateCheck
	// ImplementStage: real worker dispatch; nil = not configured.
	ImplementStage func(cfg RunConfig, plan ImplementationPlan, log RunLog) (ImplementResult, error)
	// RunGateG2: migration-apply gate.
	RunGateG2 func(worktreePath string, timeoutMs int) (G2G4Result, error)
	// RunGateG4: async-review gate.
	RunGateG4 func(worktreePath string) (G2G4Result, error)
	// RunGateG1: test gate.
	RunGateG1 func(worktreePath string, timeoutMs int) (G2G4Result, error)
	// PublishStage: PR publishing; nil = not configured.
	PublishStage func(cfg RunConfig, plan ImplementationPlan, impl ImplementResult) (string, error)
}

// RunPipeline mirrors runPipeline(): fetch → plan → G0 → spec-check →
// implement → G1/G2/G3/G4 → publish.
func RunPipeline(cfg RunConfig, deps PipelineDeps, log RunLog) ([]StageOutcome, error) {
	outcomes := []StageOutcome{}

	// Stage: fetch
	log.Info("fetch", fmt.Sprintf("Fetching ticket %s", cfg.TicketID), nil)
	ticket, err := deps.FetchTicket(cfg.TicketID)
	if err != nil {
		return nil, err
	}
	log.Info("fetch", fmt.Sprintf("Fetched %q", ticket.Title), []ledger.KV{{Key: "labels", Value: strings.Join(ticket.Labels, ",")}})

	// Stage: plan
	plan := scout.PlanFromTicket(ticket)
	log.Info("plan", fmt.Sprintf("Classified as %s", plan.Classification), []ledger.KV{{Key: "tasks", Value: strings.Join(plan.Tasks, ",")}})

	// Gate G0 (issue readiness): rejected tickets get actionable findings
	// posted to the tracker so credits are not burned on under-specified work.
	if deps.RunGateG0 != nil {
		g0 := deps.RunGateG0(ticket, plan.Classification)
		state := "failed"
		if g0.Skipped {
			state = "skipped"
		} else if g0.Passed {
			state = "passed"
		}
		detail := g0.Detail
		if i := strings.IndexByte(detail, '\n'); i >= 0 {
			detail = detail[:i]
		}
		log.Info("validate", fmt.Sprintf("G0 %s: %s", state, detail), []ledger.KV{
			{Key: "score", Value: g0.Score},
			{Key: "threshold", Value: g0.Threshold},
		})
		outcomes = append(outcomes, StageOutcome{Stage: "validate", Passed: g0.Passed})
		if !g0.Passed {
			log.Warn("clarify", "G0 readiness gate rejected ticket", []ledger.KV{{Key: "findings", Value: len(g0.Findings)}})
			if ticket.TrackerInternalID != "" && deps.PostTicketComment != nil {
				comment := g0.Detail
				if comment == "" {
					comment = "G0 readiness gate rejected this ticket"
				}
				if err := deps.PostTicketComment(ticket.TrackerInternalID, comment); err != nil {
					log.Warn("clarify", fmt.Sprintf("Failed to post G0 rejection comment: %s", err.Error()), nil)
				}
			}
			reason := g0.Detail
			if reason == "" {
				reason = "score below threshold"
			}
			outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: fmt.Sprintf("G0 readiness gate rejected: %s", reason)})
			return outcomes, nil
		}
	}

	// Spec check (FR-TICKET-05)
	spec := scout.CheckSpec(ticket)
	if !spec.Sufficient && spec.Question != "" {
		log.Warn("clarify", "Insufficient specification", nil)
		outcomes = append(outcomes, StageOutcome{Stage: "clarify", Question: spec.Question})
		if ticket.TrackerInternalID != "" && deps.PostTicketComment != nil {
			if err := deps.PostTicketComment(ticket.TrackerInternalID, spec.Question); err != nil {
				log.Warn("clarify", fmt.Sprintf("Failed to post clarification comment: %s", err.Error()), nil)
			}
		}
		return outcomes, nil
	}
	outcomes = append(outcomes, StageOutcome{Stage: "plan", Summary: plan.Classification, Tasks: plan.Tasks})

	if cfg.DryRun {
		log.Info("plan", "Dry-run: stopping before implement", nil)
		return outcomes, nil
	}

	// Stage: implement
	if deps.ImplementStage == nil {
		outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: "no worker dispatch configured"})
		return outcomes, nil
	}
	impl, err := deps.ImplementStage(cfg, plan, log)
	if err != nil {
		return nil, err
	}
	outcomes = append(outcomes, StageOutcome{
		Stage:        "implement",
		Worker:       impl.Worker,
		WorktreePath: impl.WorktreePath,
		Branch:       impl.Branch,
		Attempts:     impl.Attempts,
		OK:           impl.OK,
		KgEvidence:   impl.KgEvidence,
	})
	if !impl.OK {
		outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: "worker failed to produce a diff within retry budget"})
		return outcomes, nil
	}

	// Stage: validate — G1 (tests) then G2/G3 (migration) then G4 (async review)
	gatePath := impl.WorktreePath
	if gatePath == "" {
		gatePath = cfg.RepoPath
	}

	if deps.RunGateG1 != nil {
		g1, err := deps.RunGateG1(gatePath, cfg.TimeoutMs)
		if err != nil {
			return nil, err
		}
		logGateResult(log, "G1", g1)
		outcomes = append(outcomes, StageOutcome{Stage: "validate", Passed: g1.Passed})
		if !g1.Passed {
			detail := g1.Detail
			if detail == "" {
				detail = "no detail"
			}
			outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: fmt.Sprintf("test gate failed: %s", detail)})
			return outcomes, nil
		}
	}

	if deps.RunGateG2 != nil && plan.Classification == "migration-required" {
		g2, err := deps.RunGateG2(gatePath, cfg.TimeoutMs)
		if err != nil {
			return nil, err
		}
		logGateResult(log, "G2", g2)
		outcomes = append(outcomes, StageOutcome{Stage: "validate", Passed: g2.Passed})
		if !g2.Passed {
			detail := g2.Detail
			if detail == "" {
				detail = "no detail"
			}
			outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: fmt.Sprintf("migration apply gate failed: %s", detail)})
			return outcomes, nil
		}
	}

	if deps.RunGateG3 != nil {
		g3 := deps.RunGateG3(gatePath, plan.Classification)
		detail := g3.Detail
		if detail == "" {
			detail = ""
		}
		log.Info("validate", fmt.Sprintf("G3 %t: %s", g3.Passed, detail), nil)
		outcomes = append(outcomes, StageOutcome{Stage: "validate", Passed: g3.Passed})
		if !g3.Passed {
			outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: "migration static gate failed"})
			return outcomes, nil
		}
	}

	if deps.RunGateG4 != nil && plan.Classification == "consumer-only" {
		g4, err := deps.RunGateG4(gatePath)
		if err != nil {
			return nil, err
		}
		log.Info("validate", fmt.Sprintf("G4 %t: %s", g4.Passed, g4.Detail), nil)
		outcomes = append(outcomes, StageOutcome{Stage: "validate", Passed: g4.Passed})
		if !g4.Passed {
			detail := g4.Detail
			if detail == "" {
				detail = "no detail"
			}
			outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: fmt.Sprintf("async review gate failed: %s", detail)})
			return outcomes, nil
		}
	}

	// Stage: publish — headless mode publishes; interactive mode defers
	if cfg.AutoPr && deps.PublishStage != nil {
		prURL, err := deps.PublishStage(cfg, plan, impl)
		if err != nil {
			log.Error("publish", fmt.Sprintf("PR creation failed: %s", err.Error()), nil)
			outcomes = append(outcomes, StageOutcome{Stage: "failed", Reason: fmt.Sprintf("PR creation failed: %s", err.Error())})
			return outcomes, nil
		}
		log.Info("publish", fmt.Sprintf("PR opened: %s", prURL), nil)
		outcomes = append(outcomes, StageOutcome{Stage: "publish", PRURL: prURL, Note: "auto-pr enabled"})
	} else {
		note := "awaiting approval"
		if cfg.AutoPr {
			note = "no publisher configured"
		}
		outcomes = append(outcomes, StageOutcome{Stage: "publish", Note: note})
	}
	return outcomes, nil
}

// logGateResult renders the shared "G1 skipped|passed|failed[: first-line]"
// validate log line (pipeline.ts:151/161 shape).
func logGateResult(log RunLog, name string, g G2G4Result) {
	state := "failed"
	if g.Skipped {
		state = "skipped"
	} else if g.Passed {
		state = "passed"
	}
	detail := ""
	if g.Detail != "" {
		detail = fmt.Sprintf(": %s", firstLine(g.Detail))
	}
	log.Info("validate", fmt.Sprintf("%s %s%s", name, state, detail), nil)
}

func firstLine(s string) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		return s[:i]
	}
	return s
}
